doc/pbh/pbh-design.md

```diff
@@ -0,0 +1,758 @@
+```cpp
+bool AclTable::create()
+{
+    ...
+
+    if (type == ACL_TABLE_PBH)
+    {
+        attr.id = SAI_ACL_TABLE_ATTR_FIELD_GRE_KEY;
+        attr.value.booldata = true;
+        table_attrs.push_back(attr);
+
+        attr.id = SAI_ACL_TABLE_ATTR_FIELD_IP_PROTOCOL;
+        attr.value.booldata = true;
+        table_attrs.push_back(attr);
+
+        attr.id = SAI_ACL_TABLE_ATTR_FIELD_L4_DST_PORT;
+        attr.value.booldata = true;
+        table_attrs.push_back(attr);
+
+        attr.id = SAI_ACL_TABLE_ATTR_FIELD_INNER_ETHER_TYPE;
+        attr.value.booldata = true;
+        table_attrs.push_back(attr);
+
+        attr.id = SAI_ACL_TABLE_ATTR_ACL_STAGE;
+        attr.value.s32 = SAI_ACL_STAGE_INGRESS;
+        table_attrs.push_back(attr);
+
+        sai_status_t status = sai_acl_api->create_acl_table(&m_oid, gSwitchId, (uint32_t)table_attrs.size(), table_attrs.data());
+
+        if (status == SAI_STATUS_SUCCESS)
+        {
+            gCrmOrch->incCrmAclUsedCounter(CrmResourceType::CRM_ACL_TABLE, acl_stage, SAI_ACL_BIND_POINT_TYPE_PORT);
+            gCrmOrch->incCrmAclUsedCounter(CrmResourceType::CRM_ACL_TABLE, acl_stage, SAI_ACL_BIND_POINT_TYPE_LAG);
+        }
+
+        return status == SAI_STATUS_SUCCESS;
+    }
+
+    ...
+}
+
+class AclRulePbh: public AclRule
+{
+public:
+    AclRulePbh(AclOrch *m_pAclOrch, string rule, string table, acl_table_type_t type, bool createCounter = false);
+
+    bool validateAddAction(string attr_name, string attr_value);
+    bool validateAddMatch(string attr_name, string attr_value);
+    bool validate();
+    void update(SubjectType, void *);
+};
+```
+
+## 2.4 DB schema
+
+### 2.4.1 Config DB
+
+#### 2.4.1.1 PBH table
+```abnf
+; defines schema for PBH table configuration attributes
+key = PBH_TABLE|table_name ; table name. Must be unique
+
+; field     = value
+port_list   = port-list ; ports to which this table is applied. Can be empty
+lag_list    = lag-list  ; portchannels to which this table is applied. Can be empty
+description = *255VCHAR ; table description. Can be empty
+
+; value annotations
+port-name = 1*64VCHAR                         ; name of the port
+port-list = port-name [ 1*( "," port-name ) ] ; list of the ports. Valid values range is platform dependent
+lag-name  = "PortChannel" 1*4DIGIT            ; name of the portchannel
+lag-list  = lag-name [ 1*( "," lag-name ) ]   ; list of the portchannels. Valid values range is platform dependent
+```
+
+**Note:** at least one member of _port_list_ or _lag_list_ is required
+
+#### 2.4.1.2 PBH rule
+```abnf
+; defines schema for PBH rule configuration attributes
+key = PBH_RULE|table_name|rule_name ; rule name. Must be unique across the table
+
+; field           = value
+priority         = 1*5DIGIT      ; rule priority. Valid values range is platform dependent
+gre_key          = h32 "/" h32   ; GRE key (32 bits)
+ip_protocol      = h8 "/" h8     ; IP protocol (8 bits)
+ipv6_next_header = h8 "/" h8     ; IPv6 Next Header (8 bits)
```

Comment:
I assume, ip_protocol and ipv6_next_header fields cant co-exist, right? if no, can we just have one field as IP protocol numbers are unique? also, IP protocol numbers are not bits friendly, not sure how 0xff mask will be helpful, having an example for more than one protocol number would be useful.

IP_PROTOCOL: 0x11/0xff

Reply:
Since we use inner header names, maybe having outer name in IP protocol/gre_key..etc field descriptions would be helpful e.g Outer IP header protocol

Reply:
@venkatmahalingam we are following the SAI spec, which claims that `sai_acl_field_data_t` should have a mask. The goal is to provided user as much flexibility as possible regardless to what is actually supported by platform:
https://github.com/opencomputeproject/SAI/blob/v1.8.1/inc/saiacl.h#L1829
```
    /**
     * @brief IP Protocol
     *
     * @type sai_acl_field_data_t sai_uint8_t
     * @flags CREATE_AND_SET
     * @default disabled
     */
    SAI_ACL_ENTRY_ATTR_FIELD_IP_PROTOCOL,
```
https://github.com/opencomputeproject/SAI/blob/v1.8.1/inc/saitypes.h#L549
```
/**
 * @brief Defines a single ACL filter
 *
 * @note IPv4 and IPv6 Address expected in Network Byte Order
 *
 * @extraparam const sai_attr_metadata_t *meta
 */
typedef struct _sai_acl_field_data_t
{
    /**
     * @brief Match enable/disable
     */
    bool enable;

    /**
     * @brief Field match mask
     *
     * @note Nothing can be serialized if mask is not needed for data items
     * like object id.
     *
     * @passparam meta
     * @validonly enable == true
     */
    sai_acl_field_data_mask_t mask;

    /**
     * @brief Expected AND result using match mask above with packet field
     * value where applicable.
     *
     * @passparam meta
     * @validonly enable == true
     */
    sai_acl_field_data_data_t data;
} sai_acl_field_data_t;
```
https://github.com/Azure/sonic-swss/blob/master/orchagent/aclorch.cpp#L323
```
else if (attr_name == MATCH_IP_PROTOCOL || attr_name == MATCH_NEXT_HEADER)
{
    value.aclfield.data.u8 = to_uint<uint8_t>(attr_value);
    value.aclfield.mask.u8 = 0xFF;
}
```
The IP proto and IPv6 Next Header can't co-exist at the same level (e.g., outer/inner frame). These are two different fields.
For the attributes description (e.g., Outer IP header protocol) we have `2.2 SAI API` section. Also to distinguish between outer/inner area we are following SAI name convention:
```
SAI_ACL_ENTRY_ATTR_FIELD_IP_PROTOCOL - outer field
SAI_ACL_ENTRY_ATTR_FIELD_INNER_ETHER_TYPE - inner field
```
If you still believe that we need to add it to the schema description explicitly - no issues i can do it.

Reply:
ok, it is possible to configure more than one IP protocol with mask?

Reply:
@venkatmahalingam no. The match/action fields should be unique. this means that `ip_protocol` can appear only once.

Reply:
ok.